win32/tests/job_object_helper_int/job_object_helper_int.c

```diff
@@ -92,6 +99,235 @@ TEST_FUNCTION_CLEANUP(cleanup)
+    // call job_object_helper_set_job_limits_to_current_process
+    STARTUPINFOA si[NUM_TEST_PROCESSES];
+    PROCESS_INFORMATION pi[NUM_TEST_PROCESSES];
+
+    for (int i = 0; i < NUM_TEST_PROCESSES; ++i)
+    {
+        LogInfo("Starting process %d", i);
+        ZeroMemory(&si[i], sizeof(si[i]));
+        si[i].cb = sizeof(si[i]);
+        ZeroMemory(&pi[i], sizeof(pi[i]));
+
+        char cmdLine[512];
+        (void)snprintf(cmdLine, sizeof(cmdLine), "\"%s\" %s %d %d", fullPath, job_name, 50, 50);
+
+        /* Start the process; this process calls job_object_helper_set_job_limits_to_current_process */
+        (void)CreateProcessA(NULL, cmdLine, NULL, NULL, FALSE, DETACHED_PROCESS, NULL, NULL, &si[i], &pi[i]);
+
+        /* sleep for 2 seconds to allow the child process to run */
+        Sleep(2000);
+        /* try for 3 times before declaring a failure, as it may take time for a new process to get scheduled and create the job object  */
+        HANDLE job_object = NULL;
+        for (int j = 0; j < NUM_RETRY; ++j)
+        {
+            job_object = OpenJobObjectA(JOB_OBJECT_QUERY, FALSE, job_name);
+            if (job_object != NULL)
+            {
+                break;
+            }
+            LogInfo("OpenJobObjectA failed, will retry %d", j + 1);
+            /* sleep for 2 seconds */
+            Sleep(2000);
+        }
+
+        ASSERT_IS_NOT_NULL(job_object);
+        ASSERT_IS_TRUE(job_object != NULL, "Failed to open job object");
+
+        /* Query the Job Object to check if it has all processes associated with it */
+        DWORD bufferSize = sizeof(JOBOBJECT_BASIC_PROCESS_ID_LIST) + sizeof(ULONG_PTR) * 16;
+        JOBOBJECT_BASIC_PROCESS_ID_LIST* pidList = NULL;
+        while (1)
+        {
+            pidList = (JOBOBJECT_BASIC_PROCESS_ID_LIST*)malloc(bufferSize);
+
+            if (!QueryInformationJobObject(job_object, JobObjectBasicProcessIdList, pidList, bufferSize, NULL))
+            {
+                DWORD err = GetLastError();
+                if (err == ERROR_MORE_DATA) {
+                    // Increase buffer size and try again
+                    bufferSize *= 2;
+                    free(pidList);
+                    continue;
+                }
+            }
+            break;
+        }
+
+        ASSERT_IS_NOT_NULL(pidList);
+        ASSERT_IS_TRUE(pidList->NumberOfAssignedProcesses == (unsigned long)(i + 1), "Job object should have 1 process associated with it");
+        free(pidList);
+    }
+
+    for (int i = 0; i < NUM_TEST_PROCESSES; ++i)
+    {
+        // Terminate the process; Terminate is forceful
+        (void)TerminateProcess(pi[i].hProcess, 0);
+
+        // Just to make sure, process has terminated
+        (void)WaitForSingleObject(pi[i].hProcess, INFINITE);
+
+        // Close process and thread handles.
+        (void)CloseHandle(pi[i].hProcess);
+        (void)CloseHandle(pi[i].hThread);
+    }
+}
+
```

Comment:
This test would be REALLY nice if inside `job_object_helper_tester` we would try to allocate memory.

Then we'd see that "some" of the job_object_helper_tester would be able to allocate the memory, and some would fail. 

This would mimic really well the case in production.

Reply:
Let's have this behavior just to make sure that we test as closely as the scenario in production. I want to see this change, so I will be waiting. Just ping me vigurously when I need to look again.